In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV

In [ ]:
df=pd.read_csv("insurance.csv")

In [ ]:
df


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520
...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830
1334,18,female,31.920,0,no,northeast,2205.98080
1335,18,female,36.850,0,no,southeast,1629.83350
1336,21,female,25.800,0,no,southwest,2007.94500


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


In [ ]:
df=pd.get_dummies(df,columns=["sex","smoker","region"],drop_first=True)
boolean_cols = df.select_dtypes(include=['bool']).columns
df[boolean_cols] = df[boolean_cols].astype(int)

In [ ]:
df.head()

,age,bmi,children,charges,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,16884.92400,0,1,0,0,1
1,18,33.770,1,1725.55230,1,0,0,1,0
2,28,33.000,3,4449.46200,1,0,0,1,0
3,33,22.705,0,21984.47061,1,0,1,0,0
4,32,28.880,0,3866.85520,1,0,1,0,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age               1338 non-null   int64  
 1   bmi               1338 non-null   float64
 2   children          1338 non-null   int64  
 3   charges           1338 non-null   float64
 4   sex_male          1338 non-null   int64  
 5   smoker_yes        1338 non-null   int64  
 6   region_northwest  1338 non-null   int64  
 7   region_southeast  1338 non-null   int64  
 8   region_southwest  1338 non-null   int64  
dtypes: float64(2), int64(7)
memory usage: 94.2 KB


In [ ]:
X = df.drop('charges', axis=1)
y = df['charges']


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
def evaluate_model(y_true, y_pred):
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_true, y_pred)

    print(f"Mean Absolute Error: {mae:.2f}")
    print(f"Mean Squared Error: {mse:.2f}")
    print(f"Root Mean Squared Error: {rmse:.2f}")
    print(f"R-squared: {r2:.2f}")

Train a Linear Regression model.

In [ ]:
from sklearn.linear_model import LinearRegression


lr=LinearRegression()
lr.fit(X_train, y_train)

y_pred=lr.predict(X_test)
print("Linear Regression Model Performance:")
evaluate_model(y_test, y_pred)

Linear Regression Model Performance:
Mean Absolute Error: 4181.19
Mean Squared Error: 33596915.85
Root Mean Squared Error: 5796.28
R-squared: 0.78


Train an SGD Regressor model.

In [ ]:
from sklearn.linear_model import SGDRegressor


sgd= SGDRegressor(random_state=42)
sgd.fit(X_train, y_train)
y_pred=sgd.predict(X_test)


print("SGD Regressor Model Performance:")
evaluate_model(y_test, y_pred)

SGD Regressor Model Performance:
Mean Absolute Error: 5047835588.05
Mean Squared Error: 37187938326119555072.00
Root Mean Squared Error: 6098191397.96
R-squared: -239537791830.11


Train an SGD Regressor model with proper feature scaling.

In [ ]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

sgd_scaled = SGDRegressor(random_state=42)
sgd_scaled.fit(X_train_scaled, y_train)


y_pred= sgd_scaled.predict(X_test_scaled)

print("SGD Regressor Model Performance (with scaled features):")
evaluate_model(y_test, y_pred)

SGD Regressor Model Performance (with scaled features):
Mean Absolute Error: 4172.48
Mean Squared Error: 33608943.92
Root Mean Squared Error: 5797.32
R-squared: 0.78


Train a Decision Tree Regressor model.

In [ ]:
from sklearn.tree import DecisionTreeRegressor

dt=DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred=dt.predict(X_test)


print("Decision Tree Regressor Model Performance:")
evaluate_model(y_test, y_pred)

Decision Tree Regressor Model Performance:
Mean Absolute Error: 3195.11
Mean Squared Error: 42446908.01
Root Mean Squared Error: 6515.13
R-squared: 0.73


Train a Random Forest Regressor model without hyperparameter tuning.

In [ ]:
from sklearn.ensemble import RandomForestRegressor


rf=RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)
y_pred=rf.predict(X_test)

print("Random Forest Regressor Model Performance:")
evaluate_model(y_test, y_pred)

Random Forest Regressor Model Performance:
Mean Absolute Error: 2550.08
Mean Squared Error: 20942520.92
Root Mean Squared Error: 4576.30
R-squared: 0.87


Use `GridSearchCV` for hyperparameters tuning for Random Forest

In [ ]:
param_grid_rf = {
    'n_estimators': [50, 100, 75, 150],
    'max_depth': [3, 1, 5],
    'min_samples_split': [2, 5, 7],
    'min_samples_leaf': [1, 2, 4]
}


rf_model_grid = RandomForestRegressor(random_state=42)


grid_search_rf = GridSearchCV(
    estimator=rf_model_grid,
    param_grid=param_grid_rf,
    scoring='r2',
    cv=2,
    n_jobs=-1,
)


grid_search_rf.fit(X_train, y_train)
print("Best parameters found:", grid_search_rf.best_params_)

best_rf_model = grid_search_rf.best_estimator_


y_pred= best_rf_model.predict(X_test)
print("Tuned Random Forest Regressor Model Performance:")
evaluate_model(y_test, y_pred)

Best parameters found: {'max_depth': 3, 'min_samples_leaf': 4, 'min_samples_split': 2, 'n_estimators': 100}
Tuned Random Forest Regressor Model Performance:
Mean Absolute Error: 2726.62
Mean Squared Error: 20705711.58
Root Mean Squared Error: 4550.35
R-squared: 0.87


Use `GridSearchCV` for hyperparameters for our XGBoost model.

In [ ]:
import xgboost as xgb


xgb_model = xgb.XGBRegressor()

param_grid = {
    'n_estimators': [100, 50, 150],
    'learning_rate': [0.01, 0.1, 0.05],
    'max_depth': [3, 5, 7],
    'subsample': [0.7, 0.8, 0.9],
    'colsample_bytree': [0.7, 0.8, 0.9]
}


grid_search = GridSearchCV(
    estimator=xgb_model,
    param_grid=param_grid,
    scoring='r2',
    cv=2,
    n_jobs=-1,
)


grid_search.fit(X_train, y_train)

print("Best parameters found:", grid_search.best_params_)


best_xgb_model = grid_search.best_estimator_


y_pred_tuned_xgb = best_xgb_model.predict(X_test)
print("Tuned XGBoost Regressor Model Performance:")
evaluate_model(y_test, y_pred_tuned_xgb)

Best parameters found: {'colsample_bytree': 0.8, 'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'subsample': 0.7}
Tuned XGBoost Regressor Model Performance:
Mean Absolute Error: 2557.75
Mean Squared Error: 18823115.08
Root Mean Squared Error: 4338.56
R-squared: 0.88


Train an AdaBoost Regressor model with hyperparameter tuning.

In [ ]:
from sklearn.ensemble import AdaBoostRegressor

param_grid_adaboost = {
    'n_estimators': [50, 100, 150],
    'learning_rate': [0.01, 0.1, 0.05, 1.0],
    'loss': ['linear', 'square', 'exponential']
}

adaboost_model = AdaBoostRegressor(estimator=DecisionTreeRegressor(max_depth=3), random_state=42)


grid_search_adaboost = GridSearchCV(
    estimator=adaboost_model,
    param_grid=param_grid_adaboost,
    scoring='r2',
    cv=2,
    n_jobs=-1,
)


grid_search_adaboost.fit(X_train, y_train)


print("Best parameters found:", grid_search_adaboost.best_params_)


best_adaboost_model = grid_search_adaboost.best_estimator_


y_pred_adaboost = best_adaboost_model.predict(X_test)
print("Tuned AdaBoost Regressor Model Performance:")
evaluate_model(y_test, y_pred_adaboost)

*   XGBoost (r2 score: 0.88)
*   Random Forest(r2 score: 0.87)
*   Ada Boost(r2 score: 0.86)
*   Linear Regression(r2 score: 0.78)
*   SGD Regressor(r2 score: 0.78)
*   Descision Tree(r2 score: 0.73)

XGBoost scored the best among all the other models

As the data is non linear, linear regression and SGD regressor are not performing well.

Descision tree is overfitting hence not performing well on the training data.

XGBoost is built on the gradient boosting framework, where new models are trained to predict the residuals errors of prior models.

Also it in coorporates the regularization terms.

Random Forest performed quite well but XGBoost slightly edged it out due to the advantages of its boosting approach and explicit regularization.

AdaBoost is also a boosting algorithm, but XGBoost's advanced optimization and regularization techniques typically give it an edge in performance.







